src/core/lp_lexer.ml

```diff
@@ -0,0 +1,216 @@
+                   | math | other_math | subscript | superscript
+                   | supplemental_punctuation ]
+  let id = [%sedlex.regexp? (letter | '_'), Star (letter | digit | '_')]
+  let escid =
+    [%sedlex.regexp? "{|", Star (Compl '|' | '|', Compl '}'), Star '|', "|}"]
+  let stringlit = [%sedlex.regexp? '"', Star (Compl ('"' | '\n')), '"']
+  let float = [%sedlex.regexp? integer, '.', Opt (integer)]
+  let comment = [%sedlex.regexp? "//", Star (Compl ('\n' | '\r'))]
+
+  (** [nom buf] eats whitespaces and comments in buffer [buf]. *)
+  let rec nom : lexbuf -> unit = fun buf ->
+    match%sedlex buf with
+    | ' ' -> nom buf
+    | '\t' -> nom buf
+    | '\n' -> nom buf
+    | '\r' -> nom buf
+    | "\r\n" -> nom buf
+    | "/*" -> nom_comment buf
+    | comment -> nom buf
+    | _ -> ()
+and nom_comment : lexbuf -> unit = fun buf ->
+    match%sedlex buf with
+    | eof -> raise (SyntaxError (Pos.none "Unterminated comment."))
+    | "*/" -> nom buf
+    | any -> nom_comment buf
+    | _ -> assert false
+
+  (** [is_escaped s] is true if string [s] is escaped: i.e. enclosed in "{|
+      |}". *)
+  let is_escaped : string -> bool = fun s ->
+    let buf = Utf8.from_string s in
+    match%sedlex buf with
+    | escid -> true
+    | _ -> false
+
+  (** [is_identifier s] is true if [s] is an identifier. *)
+  let is_identifier : string -> bool = fun s ->
+    let lexbuf = Sedlexing.Utf8.from_string s in
+    match%sedlex lexbuf with
+    | id -> true
+    | _ -> false
+
+  (** [is_debug_flag s] is true if [s] is a debug flag of the form [+...] or
+      [-...] where the dots are a sequence of letters. *)
+  let is_debug_flag : string -> bool = fun s ->
+    let lexbuf = Utf8.from_string s in
+    match%sedlex lexbuf with
+    | ('+' | '-'), Plus lowercase -> true
+    | _ -> false
+
+  let token buf =
+    nom buf;
+    match%sedlex buf with
+    | eof -> EOF
+    | '(' -> L_PAREN
+    | ')' -> R_PAREN
+    | '[' -> L_SQ_BRACKET
+    | ']' -> R_SQ_BRACKET
+    | '{' -> L_CU_BRACKET
+    | '}' -> R_CU_BRACKET
+    | '_' -> WILD
+    | ':' -> COLON
+    | ',' -> COMMA
+    | '.' -> DOT
+    | ';' -> SEMICOLON
+    | '@' -> AT
+    | '?' -> QUESTION_MARK
+    | '$' -> DOLLAR
+    | '`' -> BACKQUOTE
+    | '|' -> VBAR
+    | '+', Plus alphabet ->
+        DEBUG_FLAGS(true, Utf8.sub_lexeme buf 1 (lexeme_length buf - 1))
+    | '-', Plus alphabet ->
+        DEBUG_FLAGS(false, Utf8.sub_lexeme buf 1 (lexeme_length buf - 1))
+    | 0x2254 (* ≔ *) -> ASSIGN
+    | 0x21aa (* ↪ *) -> REWRITE
+    | 0x2192 (* → *) -> ARROW
+    | 0x03bb (* λ *) -> LAMBDA
+    | 0x03a0 (* Π *) -> PI
+    | 0x2261 (* ≡ *) -> EQUIV
+    | 0x22a2 (* ⊢ *) -> TURNSTILE
+    | "prover_timeout" -> PROVER_TIMEOUT
+    | "reflexivity" -> REFL
+    | "inductive" -> INDUCTIVE
+    | "symmetry" -> SYMMETRY
+    | "rewrite" -> REWRITE
+    | "assume" -> INTRO
+    | "refine" -> REFINE
+    | "simpl" -> SIMPL
+    | "apply" -> APPLY
+    | "proof" -> PROOF
```

Comment:
to be replaced by "begin"
side remark: thanks to put all these strings in alphabetical order